### DISCUSSION TITLE
Linearization Tool

### DISCUSSION THREAD

**Discussion by htessaro** [CURENT/andes#153] [Q&A] [answered]
Dear Dr Cui,

First, congratulation on the amazing work with ANDES.

I am wondering if ANDES offers the possibility of linearization of the used system. Is it possible to obtain a linearized system for a particular operation point after the definition of the system? Thanks.

Best regards
Heitor José Tessaro

**Comment by cuihantao** [comment, marked as answer]
Hi Heitor,

Thank you. The eigenvalue analysis routine does the Linearization.

You can call system.EIG.run() at any point after initializing the dynamics. If you want to linearization the system following a disturbance, set the tome domain simulation end time to the desired time, simulate it, and run system.EIG.run().

Regards,
Hantao Cui

**Reply by htessaro** [reply]
Thank you Dr Cui,

With the system.EIG.run() function, I would obtain the system eigenvalues. However, is it possible to get the state-space matrices from the linearized system? 

My objective is to design a centralized controller. The design method requires the state-space matrices from the linearized system. For this reason, I ask the question above.

Best regards,
Heitor

**Reply by cuihantao** [reply]
@heitortessaro , sorry for the late response. 

ANDES generates the state matrix A from the linearized system. You should be able to manually generate the B matrix by taking the derivative of your input w.r.t. the states and place the derivatives at the corresponding indices.

```
ss.EIG.run() # run the eigenvalue analysis routine
As = ss.EIG.As  # As is the state matrix A
names = ss.EIG.x_name  # the corresponding variable names 
```

Note that the variable names in `ss.EIG.x_name` is not always the same as `ss.dae.x_name`. A state variable will be moved to the end if it causes numerical singularity (associated with zero time constants, just like an algebraic variable). For details, see the code in `eig.py`.
